src/falcon/falcon_c/rpo.c

```diff
@@ -0,0 +1,757 @@
+     * of local endianness.
+     */
+    memset(rc->st.A, 0, sizeof rc->st.A);
+}
+
+void rpo128_inc_absorb(inner_rpo128_context *rc, const uint8_t *in, size_t len)
+{
+    size_t dptr;
+
+    dptr = (size_t)rc->dptr;
+    while (len > 0)
+    {
+        size_t clen, u;
+
+        /* 136 * 8 = 1088 bit for the rate portion in the case of SHAKE256
+            For RPO, this is 64 * 8 = 512 bits
+            The capacity for SHAKE256 is at the end while for RPO128 it is at the beginning */
+        clen = 96 - dptr;
+        if (clen > len)
+        {
+            clen = len;
+        }
+#if FALCON_LE // yyyLE+1
+        for (u = 0; u < clen; u++)
+        {
+            rc->st.dbuf[dptr + u] = in[u];
+        }
+#else  // yyyLE+0
+        for (u = 0; u < clen; u++)
+        {
+            rc->st.dbuf[dptr + u] = in[u];
+        }
+#endif // yyyLE-
+        dptr += clen;
+        in += clen;
+        len -= clen;
+        if (dptr == 96)
+        {
+            process_block(rc->st.A);
+            dptr = 32;
+        }
+    }
+    rc->dptr = dptr;
+}
+
+void rpo128_inc_finalize(inner_rpo128_context *rc)
+{
+    /*
+     * Set dptr to the end of the buffer, so that first call to
+     * extract will process the block.
+     */
+    rc->dptr = 96;
+}
+
+void rpo128_inc_squeeze(inner_rpo128_context *rc, uint8_t *out, size_t len)
+{
+    size_t dptr;
+
+    dptr = (size_t)rc->dptr;
+    while (len > 0)
+    {
+        size_t clen;
+
+        if (dptr == 96)
+        {
+            process_block(rc->st.A);
+            dptr = 32;
+        }
+        clen = 96 - dptr;
+        if (clen > len)
+        {
+            clen = len;
+        }
+        len -= clen;
+#if FALCON_LE // yyyLE+1
+        memcpy(out, rc->st.dbuf + dptr, clen);
+        dptr += clen;
+        out += clen;
+#else  // yyyLE+0
+        memcpy(out, rc->st.dbuf + dptr, clen);
+        dptr += clen;
+        out += clen;
+#endif // yyyLE-
+    }
+    rc->dptr = dptr;
+}
+
+void PQCLEAN_FALCON512_CLEAN_hash_to_point_rpo(
+    inner_rpo128_context *rc,
+    uint16_t *x, unsigned logn)
+{
+    /*
+     * This implementation avoids the rejection sampling step needed in the
+     * per-the-spec implementation. It uses a remark in https://falcon-sign.info/falcon.pdf 
+     * page 31, which argues that the current variant is secure for the parameters set by NIST.
+     * Avoiding the rejection-sampling step leads to an implementation that is constant-time.
+     * TODO: Check that the current implementation is indeed constant-time.
+     */
+    size_t n;
+
+    n = (size_t)1 << logn;
+    while (n > 0)
+    {
+
+        uint8_t buf[8];
+        uint64_t w;
+
+        inner_rpo128_extract(rc, (void *)buf, sizeof buf);
+        w = ((uint64_t)(buf[7]) << 56) | ((uint64_t)(buf[6]) << 48) | ((uint64_t)(buf[5]) << 40) | ((uint64_t)(buf[4]) << 32) | ((uint64_t)(buf[3]) << 24) | ((uint64_t)(buf[2]) << 16) | ((uint64_t)(buf[1]) << 8) | ((uint64_t)(buf[0]));
+
+        w %= 12289;
```

Comment:
Should we have a constant defined for $12289$?

Reply:
Done!